### DISCUSSION TITLE
Pushing Ragna UI to its limits

### DISCUSSION THREAD

**Discussion by Tengal-Teemo** [Quansight/ragna#239] [General]
# Pushing Ragna UI to its limits
I thought it would be a good idea to try limit testing Ragna within what would be considered its reasonable use-case.  I have tested some limits that are significant to my work, and come up with some interesting findings and potential solutions.

## System:
 - OS: Linux Mint, Victoria (21.2)
 - Python Version: 3.10.12
 - Ragna Version: v0.1.1 (These issues still exist in v0.1.2rc0 however)
 - Panel Version: 1.3.4
 - CPU: Intel I5 11600-K 

For these investigations, I added some rudimentary logging to both Ragna and Panel using the python library timeit.

## Large numbers of documents
The only issue with this has already been addressed in #224, but I'd just like to quantify the performance improvement.  When loading a corpus of 2224 documents, with the pills header the upload took roughly 221.54 seconds.  
```
Updated Model: 0.1184683400006179
Setting Current Chat...
Creating RagnaChatInterface...
Recreating Creating Chat Messages...
Got chat messages in 0.0355644970004505
Created chat interface in 0.07898727900101221
Updated Model: 0.14176178999878175
Set Current Chat in 221.5431228969992
```

Patching out the pills header and performing the same upload yielded 0.15 seconds, quite a large improvement.

```
Updated Model: 0.019271736000519013
Setting Current Chat...
Creating RagnaChatInterface...
Recreating Creating Chat Messages...
Got chat messages in 0.03131087099973229
Created chat interface in 0.07647951199942327
Updated Model: 0.04566194499966514
Set Current Chat in 0.14838966999923286
```

The primary cause of this appears to be the way that panel implements loading.  The 64 bit encoded SVG is uploaded for every single element on that page, including all pills and all chat messages.  This was observed with tcpflow producing the following logs from the command `sudo tcpflow -p -c -i any port 31477 | grep "pn-dots"`

```
27.000.000.001.31477-127.000.000.001.41776: ..{}..........{"events":[{"kind":"ModelChanged","model":{"id":"p1049"},"attr":"children","new":[{"type":"object","name":"panel.models.layout.Card","id":"p1929","attributes":{"name":"Card02344","css_classes":["chat-interface"],"styles":{"type":"map","entries":[["padding","0px"]]},"stylesheets":["\n:host(.pn-loading.pn-dots):before, .pn-loading.pn-dots:before {\n  background-image: url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL
```
This is not the full message.  The string `PHN2ZyB4bWxucz0iaHR0cDovL` is the beginning of the SVG for the loading icon.  This is sent for every single element on the page as observed through TCP flow.  This is more of a bug in panel than ragna, and unless the page is composed of thousands of elements poses no issue.

Note: Quite strange is that the long setting of the chat only appeared **after the first upload** of the corpus.  When uploaded for the first time, the loading time was noticeably shorter (only 25 seconds) and the logs show the current chat set in only 5.6 seconds.
```
Updated Model: 0.010828394999407465
Setting Current Chat...
Updated Model: 2.4782531910004764
Creating RagnaChatInterface...
Recreating Creating Chat Messages...
Got chat messages in 0.5006771669995942
Created chat interface in 0.545532206999269
Updated Model: 0.00013107199993100949
Updated Model: 0.1688477759998932
Set Current Chat in 5.660886992998712
```
For the remaining tests and experiments, I patched out the pills by commenting out lines 596-623 in ragna/_ui/central_view.py

## Large numbers of chat messages
When switching between chats with a large number of messages, there is a very large delay.  The main cause is determined to be from 2 places.  The first is the re-construction of the list of RagnaChatMessages.  The second is the re-rendering of the bokeh model after the update.  See below for logs from swapping between 2 chats (100 Questions, 100 Responses, 2 character query, 2700 character response)
```
Setting Current Chat...
Creating RagnaChatInterface...
Recreating Creating Chat Messages...
Got chat messages in 6.884512655000435
Updated Model: 4.999417820999952
Set Current Chat in 12.58346340999924
```
### Reconstruction of the list of RagnaChatMessages
This takes a long time because ragna is built from panel.  Analysing this further, almost the entire time constructing the RagnaChatMessage comes from constructing the underlying panel.chat.ChatMessage object underneath.  The best solution for this is to cache the RagnaChatMessage and ChatMessage objects when entering and exiting chats, or upon generation.  With this implemented, performance is improved:
```
Got chat entries in 3.0379997042473406e-06 seconds
Updated Model: 2.9470985920015664
Set current chat in 3.4143335570006457
```
The remaining issue is the update time.  This also appears to be an underlying issue with panel, as when the chat interface is recreated the bokeh model needs to be re-rendered with a massive number of changes.  This was verified by recreating a base case using panel and a similar number of chat messages.  This was verified by porting RagnaChatMessages to a separate panel app and testing them against the most basic panel.chat.ChatMessage objects.
```
// Ragna Chat Messages
Updated model in 1.7454483669998808
2023-12-08 09:23:17,897 WebSocket connection opened
2023-12-08 09:23:17,897 ServerConnection created
// Generic Chat Messages
Updated model in 1.7793840659996931
2023-12-08 09:24:30,838 WebSocket connection opened
2023-12-08 09:24:30,839 ServerConnection created
```
In my opinion the best way to fix this would be to implement central_view inside of panel tabs, and tie each tab to a chat id.  This means that the bokeh model is never re-rendered with a lot of objects at once (provided the tabs aren't set to dynamic).

## Large numbers of chats
There appears to be a bug with having 25 or more chats.  You cannot have more than 25 chats.  Upon creation of the 25th chat, the modal will say "an error occurred".  Clicking "Start Conversation" again will result in the creation of the chat, but any more presses of the + New Chat button will result in the following error message:
```
ERROR:bokeh.server.protocol_handler:error handling message
 message: Message 'PATCH-DOC' content: {'events': [{'kind': 'MessageSent', 'msg_type': 'bokeh_event', 'msg_data': {'type': 'event', 'name': 'button_click', 'values': {'type': 'map', 'entries': [['model', {'id': 'p16815'}]]}}}]} 
 error: UnknownReferenceError("can't resolve reference 'p16815'")
Traceback (most recent call last):
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/server/protocol_handler.py", line 97, in handle
    work = await handler(message, connection)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/server/session.py", line 94, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/server/session.py", line 286, in _handle_patch
    message.apply_to_document(self.document, self)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/protocol/messages/patch_doc.py", line 104, in apply_to_document
    invoke_with_curdoc(doc, lambda: doc.apply_json_patch(self.payload, setter=setter))
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/document/callbacks.py", line 443, in invoke_with_curdoc
    return f()
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/protocol/messages/patch_doc.py", line 104, in <lambda>
    invoke_with_curdoc(doc, lambda: doc.apply_json_patch(self.payload, setter=setter))
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/document/document.py", line 374, in apply_json_patch
    patch: PatchJson = deserializer.deserialize(patch_json)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 510, in deserialize
    return self.decode(obj.content, obj.buffers)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 525, in decode
    return self._decode(obj)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 566, in _decode
    return {key: self._decode(val) for key, val in obj.items()}
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 566, in <dictcomp>
    return {key: self._decode(val) for key, val in obj.items()}
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 568, in _decode
    return [self._decode(entry) for entry in obj]
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 568, in <listcomp>
    return [self._decode(entry) for entry in obj]
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 566, in _decode
    return {key: self._decode(val) for key, val in obj.items()}
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 566, in <dictcomp>
    return {key: self._decode(val) for key, val in obj.items()}
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 535, in _decode
    return self._decoders[type](obj, self)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/events.py", line 186, in from_serializable
    decoded_values = decoder.decode(values)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 520, in decode
    return self._decode(obj)
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 547, in _decode
    return self._decode_map(cast(MapRep, obj))
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 598, in _decode_map
    return { self._decode(key): self._decode(val) for key, val in entries }
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 598, in <dictcomp>
    return { self._decode(key): self._decode(val) for key, val in entries }
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 564, in _decode
    return self._decode_ref(cast(Ref, obj))
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 578, in _decode_ref
    self.error(UnknownReferenceError(id))
  File "/home/tazukiswift/miniconda3/envs/ragna-dev/lib/python3.9/site-packages/bokeh/core/serialization.py", line 729, in error
    raise error
bokeh.core.serialization.UnknownReferenceError: can't resolve reference 'p16815'
```
I will be creating an issue describing this bug properly. Edit: this issue has been created here #240 

## Overall Consensus
I think that the method of repeatedly reconstructing the chats needs to be removed.  At least, some kind of caching of the chat_message objects should be done on the server side so they don't have to be recreated every time you change chats.  Ultimately even with all the changes and patches suggested here, it would still take a very long time to load all chats on refresh, speaking to the issues with the underlying panel library.  Perhaps a more performant library could be used, but I have not done any research as to what or making any kind of comparison.  I'd be happy to help with any changes that are made as an outcome of this post as well.  Thanks for reading this long discussion post :smile:

**Comment by ahuang11** [comment]
Thanks for providing these insights!

I believe large numbers of ChatMessages can be addressed by adding pagination. I created an issue here: https://github.com/holoviz/panel/issues/6021

I'll also see whether ChatMessage can be optimized to render a bit quicker.

**Comment by pmeier** [comment]
Thanks a ton for this in-depth post @Tengal-Teemo! I'll convert the actionable items here into into issues for Ragna and Panel.

> ## Large numbers of documents

Apart from the actual SVG loading, we are using a [`panel.Row`](https://panel.holoviz.org/reference/layouts/Row.html) as container for the pills: 

https://github.com/Quansight/ragna/blob/7b176528c7524edc8bcf8754695caa5a363a8e23/ragna/deploy/_ui/central_view.py#L632

I don't think there is an option to automatically hide items based on width. #224 just limited the maximum number we are going to create, but ultimately this should be dynamic.

> In my opinion the best way to fix this would be to implement central_view inside of panel tabs, and tie each tab to a chat id. This means that the bokeh model is never re-rendered with a lot of objects at once (provided the tabs aren't set to dynamic).

I've advocated for caching from the beginning, but we left it out for simplicity in the first release. So I agree this is needed. If we assume that each user only has at most one session, implementing caching should be fairly straight forward.

**Reply by Tengal-Teemo** [reply]
I'd be happy to collaborate with you on any of these issues to help get them improved/fixed.  Particularly, I could have a look at implementing the caching if you don't have time for that.  Please let me know as I'm happy to help.

**Reply by pmeier** [reply]
Yes, that would be nice. Could you open an issue first detailing what you had in mind for us to discuss? I want to avoid you sending a PR if we are not sure about the solution yet.

**Reply by Tengal-Teemo** [reply]
Sorry for the absence, I was away at a competition.  Having returned, I can open an issue.  I am still not 100% sure as to the correct solution to all these problems, but we will discuss in the issue.

**Reply by pmeier** [reply]
I'm working on refactoring the UI internals in #261. Please have a look. I'll likely will put in the caching along the way.

**Reply by Tengal-Teemo** [reply]
Sure, taking a look now

**Reply by Tengal-Teemo** [reply]
I had a look, and I think adding caching will add a band-aid on the issue however having had some time to properly think about the problem, I think that the only real way to fix this issue is either to somehow cache the objects to the API (not sure how you would go about doing this) or to find a quicker way to construct the chat messages (this is a panel problem).  I am not that experienced with python as I am primarily an embedded developer until recently, so I'm not sure if this is a practical solution to the problem and there may be something I am missing.  Do you still want me to open the issue?

**Reply by ahuang11** [reply]
I believe the ChatMessage has been slightly optimized:
https://github.com/holoviz/panel/issues/6064

**Reply by philippjfr** [reply]
I'm working on a number of optimizations in Panel. What would be super helpful is a reproducible example as a baseline to compare against.

**Reply by Tengal-Teemo** [reply]
I'm on it right now, give me 10 minutes

**Reply by Tengal-Teemo** [reply]
@philippjfr Bootstrap the following code.  For reference, it took me 33 seconds to construct these messages, and ~2 seconds to make it servable (which involves updating the bokeh model).  I believe this is analogous to what is being done in Ragna when the messages are reconstructed.
``` Python
import panel as pn
import panel.chat
from timeit import default_timer as timer

pn.extension()

def callback(contents, user, instance):
    return "Ran the Chat Callback"

chat_interface = pn.chat.ChatInterface(callback=callback)

objects = []

start_creating_100_messages = timer()
for i in range(100):
    start = timer()
    objects.append(panel.chat.ChatMessage("lorem ipsum"))
    print("Created a ChatMessage in "+str(timer() - start))
print("Created 100 ChatMessages in "+str(timer() - start_creating_100_messages))

chat_interface.objects = objects

my_template = pn.template.FastListTemplate(
        title="DASHBOARD TITLE",
        main=[
            chat_interface
        ],
    )

start = timer()
my_template.servable()
print("Servable in "+str(timer() - start))

```

**Reply by ahuang11** [reply]
On main, I'm getting:
```
2023-12-21 13:24:13,842 Starting Bokeh server with process id: 59962
Created a ChatMessage in 0.010024915973190218
Created a ChatMessage in 0.01009100000374019
...
Created 100 ChatMessages in 1.0366324170026928
Servable in 1.4765250419732183
```

**Reply by philippjfr** [reply]
Indeed, there was a big bottleneck in ChatMessage creation, we were accidentally relying on loading icons from an external provider who turned off access which resulted in those requests timing out. That alone brought the timing down from 30s to 2.5s and then I was able to get that under a second on my machine with some smaller optimizations. All of that was released in panel 1.3.6.

**Reply by philippjfr** [reply]
Param 2.0.2, once released, should provide another 10% speedup.

**Reply by Tengal-Teemo** [reply]
Oh yes I should mention I believe that Ragna required version 1.3.4.  If it's been fixed in the new version that should fix a lot of things.  @pmeier

**Reply by pmeier** [reply]
Ragna never required 1.3.4 Any 1.3.* will do. 1.3.4 was just the latest release when you were testing.

**Reply by Tengal-Teemo** [reply]
Apologies, thank you for clearing that up

**Reply by pmeier** [reply]
No worries. For future references, here is the specification:

https://github.com/Quansight/ragna/blob/0bd014adb74c336f580723846134f4e1f5218b72/pyproject.toml#L31

Note that we just bumped the minimum required version in #259 for the `ragna==0.1.2` release.

**Comment by philippjfr** [comment]
@Tengal-Teemo Any chance you could share the code you used to profile and generate the timings?

**Reply by Tengal-Teemo** [reply]
@philippjfr Sure.  For panel, I modified the `_update_model` function from the `Panel` class in `base.py` as shown below:
``` Python
from timeit import default_timer as timer()

    def _update_model(
        self, events: Dict[str, param.parameterized.Event], msg: Dict[str, Any],
        root: Model, model: Model, doc: Document, comm: Optional[Comm]
    ) -> None:
        start = timer()
        msg = dict(msg)
        inverse = {v: k for k, v in self._property_mapping.items() if v is not None}
        preprocess = any(inverse.get(k, k) in self._preprocess_params for k in msg)

        # ALERT: Find a better way to handle this
        if 'styles' in msg and root is model and 'overflow-x' in msg['styles']:
            del msg['styles']['overflow-x']

        obj_key = self._property_mapping['objects']
        update_children = obj_key in msg
        if update_children:
            old = events['objects'].old

            children, old_children = self._get_objects(model, old, doc, root, comm)
            msg[obj_key] = children

            msg.update(self._compute_sizing_mode(
                children,
                dict(
                    sizing_mode=msg.get('sizing_mode', model.sizing_mode),
                    styles=msg.get('styles', model.styles),
                    width=msg.get('width', model.width),
                    min_width=msg.get('min_width', model.min_width),
                    margin=msg.get('margin', model.margin)
                )
            ))
        else:
            old_children = None

        with hold(doc):
            update = Panel._batch_update
            Panel._batch_update = True
            try:
                with freeze_doc(doc, model, msg, force=update_children):
                    super()._update_model(events, msg, root, model, doc, comm)
                    if update:
                        stop = timer()
                        print("Updated Model (Returning): " + str(stop - start))
                        return
                    from ..io import state
                    ref = root.ref['id']
                    if ref in state._views and preprocess:
                        state._views[ref][0]._preprocess(root, self, old_children)
            finally:
                Panel._batch_update = update
        stop = timer()
        print("Updated Model: " + str(stop-start))
```
This is all I did, further information on how to create the cases in which the timing issues persisted can be found in the original post.  Let me know if you have more questions.

Edit: clarification:  I just added the lines using timer()

**Reply by Tengal-Teemo** [reply]
For constructing the chat messages I just did this inside the constructor for RagnaChatMessage
```Python
start = timer()
super().__init__(**(params | kwargs))
print("Constructed ChatMessage in "+str(timer() - start))
```
On line 146 of `central_view.py`
